### PR TITLE
Don't lose continuation on multiple FlushAsync call

### PR DESCRIPTION
**Pull request by pakrym**
Fixes: https://github.com/dotnet/corefx/issues/34014

### CODE REVIEW

File: src/System.IO.Pipelines/tests/FlushAsyncTests.cs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+        [Fact]
+        public async Task CallingFlushAsyncMultipleTimesAllowsFirstToComplete()
+        {
+            Pipe.Writer.WriteEmpty(65);
+            var flushTask = Pipe.Writer.FlushAsync().AsTask();
+
+            Pipe.Writer.WriteEmpty(1);
+
+            _ = Pipe.Writer.FlushAsync();
```
Comment (halter73) [inline src/System.IO.Pipelines/tests/FlushAsyncTests.cs]: Can we add a test to verify that awaiting the second call to FlushAsync() results in an InvalidOperationException? I think that would have also caught this regression.

File: src/System.IO.Pipelines/src/System/IO/Pipelines/PipeAwaitable.cs [hunk 2]
```diff
@@ -86,6 +86,9 @@ private void ExtractCompletion(out CompletionData completionData)
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public void Reset()
         {
+            Debug.Assert(_completion == null);
+            Debug.Assert(_completionState == null);
+
             _completion = null;
```
Comment (halter73) [inline src/System.IO.Pipelines/src/System/IO/Pipelines/PipeAwaitable.cs]: As discussed in person, the nulling out can live solely in ExtractCompletion

File: src/System.IO.Pipelines/src/System/IO/Pipelines/PipeAwaitable.cs [hunk 3]
```diff
@@ -74,22 +74,26 @@ private void ExtractCompletion(out CompletionData completionData)
...
+            _executionContext = null;
 
             completionData = currentCompletion != null ?
-                new CompletionData(currentCompletion, currentState, _executionContext, _synchronizationContext) :
+                new CompletionData(currentCompletion, currentState, executionContext, synchronizationContext) :
                 default;
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public void Reset()
```
Comment (halter73) [inline src/System.IO.Pipelines/src/System/IO/Pipelines/PipeAwaitable.cs]: I think this might be better named SetUncompleted or something.

**Comment by davidfowl** [comment]
I’d like to tighten up our handling of concurrent operations. Right now it’s a bit racy when it happens (I’ll file an issue)

**Pull request merged**